clang/include/clang/CIR/Dialect/IR/CIRAttrs.td

```diff
@@ -248,13 +248,13 @@ def FPAttr : CIR_Attr<"FP", "fp", [TypedAttrInterface]> {
 
 def ConstPtrAttr : CIR_Attr<"ConstPtr", "ptr", [TypedAttrInterface]> {
   let summary = "Holds a constant pointer value";
-  let parameters = (ins AttributeSelfTypeParameter<"">:$type, "uint64_t":$value);
+  let parameters = (ins AttributeSelfTypeParameter<"">:$type, "int64_t":$value);
   let description = [{
     A pointer attribute is a literal attribute that represents an integral
     value of a pointer type.
   }];
   let builders = [
-    AttrBuilderWithInferredContext<(ins "Type":$type, "uint64_t":$value), [{
+    AttrBuilderWithInferredContext<(ins "Type":$type, "int64_t":$value), [{
```

Comment:
This change can be strange. A "negative" pointer value does not make any sense in general.

Reply:
I understand `ConstPtrAttr` to be equivalent to `llvm.inttoptr`. `#cir.ptr<-8>` can be translated as `ptr inttoptr (i64 -8 to ptr)`. Or should a new attribute be added to represent relative offset?

Reply:
> Intuitivelly it doesn't make much sense, but in practice we find code like that all around, simple example: https://godbolt.org/z/aMdxfvrf3. This is how CIR handles it: https://godbolt.org/z/55fx96z1Y. Note that if we #cir.ptr was a bit more flexible here we wouldn't need the unary op.

OK I can buy this.

> I'm more inclined towards (3), ... Any extra thoughts?

The problem of (3) is that a single vtable may contain multiple offset-to-top elements in the presence of multiple inheritance. For example:

```cpp
struct Left { virtual ~Left(); };
struct Right { virtual ~Right(); };
struct Derived : Left, Right {};
```

```llvm
@vtable for Derived = linkonce_odr dso_local unnamed_addr constant { [4 x ptr], [4 x ptr] } {
  [4 x ptr] [
    ptr null,                      ;; Offset-to-top from Left to Derived
    ptr @typeinfo for Derived,
    ptr @Derived::~Derived(),
    ptr @Derived::~Derived()
  ],
  [4 x ptr] [
    ptr inttoptr (i64 -8 to ptr),  ;; Offset-to-top from Right to Derived
    ptr @typeinfo for Derived,
    ptr @non-virtual thunk to Derived::~Derived(),
    ptr @non-virtual thunk to Derived::~Derived()
  ]
}, comdat, align 8
```

Thus if offset-to-top values are kept separate from `vtable_data` we may need a way to track which base class each of them corresponds to.

Reply:
> Thus if offset-to-top values are kept separate from vtable_data we may need a way to track which base class each of them corresponds to.

As long as offset-to-top is a list of the same size as vtable_data, all should be good, shouldn't it?

But more that that, we could track multiple lanes just as easy with some other approaches. The vtable right now is still a llvm friendly layout, but it could be anything that makes more sense / easier to use - as long as it could keep lowering to the same LLVM code.

Reply:
I updated this PR attempting to modify #cir.ptr to accept mlir::integer instead of plain uint64_t.

Reply:
Thanks @Laity000, but previously mentioned, changing #cir.ptr should be another PR on it own, can you please address it outside of this one?

Reply:
> In general, though not related to vtables, This could be used for things like tagging pointers with some sort of non-null sentinel, etc. The way to thing about this is that the constant initialization isn't related to the pointee. We should be able to write `#cir.ptr<-1 : i64> : !cir.ptr<whatever>`

Sorry for the delay of a few days. I've created a new PR #598 . Is it possible to directly use the int64_t type as the value of the `#cir.ptr`? If using MLIR integer types like `#cir.ptr<-1 : i64>`, I understand we should consider the address size based on XX ABI to determine the type, include `#cir.ptr<null>`. Can we just not worry about the address size when we create it, and instead get it from the `DataLayout` when we use the attribute?